smaug/python/ops/control_flow_ops.py

```diff
@@ -30,9 +30,77 @@ def merge(input_tensors, name="merge"):
       output_tensors_dims=[input_tensors[0].shape.dims],
       output_tensor_layout=input_tensors[0].shape.layout)[0]
+
+def cond(predication, true_fn, false_fn, name="cond"):
+  """A conditional operator.
+
+  This operator provides the capability of doing if-else statement. Depending on
+  the predication value, either the True or the False body of the operator will
+  be executed.
+
+  Args:
+    predication: A predication tensor of value 0 or 1, determining which path to
+      execute.
+    true_fn: The callable to be performed if `predication` is 1.
+    false_fn: The callable to be performed if `predication` is 0.
+
+  Returns:
+    The tensors returned by either true_fn or false_fn.
+  """
+
+  def _insert_switch_nodes(predication, branch, graph):
+    """Insert switch nodes for external tensros in the subgraph."""
+    nodes = [node for node in graph.get_nodes() if node.op != types_pb2.Data]
+    internal_tensors = []
+    for node in nodes:
+      internal_tensors.extend([tensor.name for tensor in node.output_tensors])
+    for node in nodes:
+      for i, tensor_proto in enumerate(node.input_tensors):
+        # If any input tensor of the graph appear in the graph workspace, then
+        # this tensor is an external to the graph and we create a switch node
+        # for it.
+        # Don't create switch node for an existing one.
+        if node.op == types_pb2.Switch:
+          continue
+        if tensor_proto.name not in internal_tensors:
+          source_node = graph.get_node(node.parents[i], True)
+          tensor = tensor_utils.from_tensor_proto(tensor_proto)
+          if source_node is not None:
+            tensor.source = (source_node, node.src_tensors_indices[i])
+          switch_result = switch(tensor, predication)[branch]
+          switch_result.to_tensor_proto(node.input_tensors[i])
+          # Update the node with the switch node as its new parent.
+          switch_node = switch_result.source[0]
+          node.parents[i] = switch_node.name
+          node.src_tensors_indices[i] = branch
+
+  cur_graph = global_vars.get_graph()
+  backend = cur_graph.backend
+  mem_policy = cur_graph.mem_policy
+  name = cur_graph.create_unique_name(name)
+
+  # Build the subgraph for the true branch.
+  with Graph(name="%s_true_branch" % name, backend=backend,
+             mem_policy=mem_policy) as subgraph_t:
+    res_t = true_fn()
+    if not isinstance(res_t, (list, tuple)):
+      res_t = [res_t]
+    _insert_switch_nodes(predication, 1, subgraph_t)
+  cur_graph.merge(subgraph_t)
+
+  # Build the subgraph for the false branch.
+  with Graph(name="%s_false_branch" % name, backend=backend,
+             mem_policy=mem_policy) as subgraph_f:
+    res_f = false_fn()
+    if not isinstance(res_f, (list, tuple)):
+      res_f = [res_f]
+    _insert_switch_nodes(predication, 0, subgraph_f)
+  cur_graph.merge(subgraph_f)
+
+  # Add the merge nodes for the outputs.
+  merges = [merge([f, t]) for (f, t) in zip(res_f, res_t)]
```

Comment:
nit: i think it's a little strange that the false branch comes before the true branch in merge ops. It's just a bit backwards from the typical line of thought. Do you agree?

Reply:
Right, changed the order of the branch results.